src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -175,223 +138,310 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
                 return;
             }
 
-            var declarationStatement = (StatementSyntax)declarator.Parent.Parent;
-
-            // If we convert this to 'if (o is Type x)' then 'x' will not be definitely assigned
-            // in the Else branch of the IfStatement, or after the IfStatement. Make sure
-            // that doesn't cause definite assignment issues.
-            if (!CheckDefiniteAssignment(semanticModel, localSymbol,
-                    declarationStatement, targetStatement, parentBlock, isNegativeNullCheck))
+            if (!CanSafelyConvertToPatternMatching())
             {
                 return;
             }
 
             // Looks good!
             var additionalLocations = ImmutableArray.Create(
-                declarationStatement.GetLocation(),
-                targetStatement.GetLocation(),
-                leftmostCondition.GetLocation(),
+                localStatement.GetLocation(),
+                comparison.GetAncestor<StatementSyntax>().GetLocation(),
+                comparison.GetLocation(),
                 asExpression.GetLocation());
 
             // Put a diagnostic with the appropriate severity on the declaration-statement itself.
             syntaxContext.ReportDiagnostic(Diagnostic.Create(
                 GetDescriptorWithSeverity(styleOption.Notification.Value),
-                declarationStatement.GetLocation(),
+                localStatement.GetLocation(),
                 additionalLocations));
-        }
-
-        // Ensure that all usages of the pattern variable are
-        // in scope and definitely assigned after replacement.
-        private static bool CheckDefiniteAssignment(
-            SemanticModel semanticModel,
-            ISymbol localVariable,
-            StatementSyntax declarationStatement,
-            StatementSyntax targetStatement,
-            BlockSyntax parentBlock,
-            bool isNegativeNullCheck)
-        {
-            var statements = parentBlock.Statements;
-            var targetIndex = statements.IndexOf(targetStatement);
-            var declarationIndex = statements.IndexOf(declarationStatement);
 
-            // Since we're going to remove this declaration-statement,
-            // we need to first ensure that it's not used up to the target statement.
-            if (declarationIndex + 1 < targetIndex)
+            bool CanSafelyConvertToPatternMatching()
```

Comment:
if this is a local function, can you have a "return" statement above it.  and a comment "// local functions".  as well?

Reply:
this function needs a strong comment explaining what it's going and what the overall intuition should be here.